[CORE-12817] - Hook kafka::client::cluster into link

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR will create an instance of the `kafka::client::cluster` class based on the configuration provided for the cluster link

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
k/c/cluster: Tracking topic authorized operations

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Updated tls settings

TLS settings can now be provided via a file path or the PEM value

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Introduce deps

Useful for dependencies

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Updated error codes

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/client/bzl: Make cluster public

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c/cl: Updated validation

Improved validation of TLS configuration

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added client_id to connection config

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Added `client_id` to model config

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e79b34b0-35c3-436c-a8c1-e5c777dc9015-->
<!--TEST RESULTS-->
<!--169299cd-baa0-427f-818c-b9b48d7155c2-->
<!--TEST RESULTS-->
<!--c635698a-b8b2-4465-aff4-1da98af6c550-->
<!--TEST RESULTS-->
<!--311fb127-c5c0-4eb0-91db-15ca2e31044e-->
<!--TEST RESULTS-->
<!--d08162e8-e8ea-42e3-bdfc-12d3ae183c78-->
<!--TEST RESULTS-->
<!--1a3b6bed-b256-45a8-ad21-008cbfb59db0-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69127">build#69127</a></summary>


| test_class               | test_method                          | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_creating_and_listing_migrations |                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69127#0198151f-1371-4f06-90bf-dc7c4d2c7035 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| DatalakeE2ETests         | test_json_schema_unicode             | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69127#0198151f-1371-4f06-90bf-dc7c4d2c7035 | FLAKY         | 19/21    | upstream reliability is '98.18181818181819'. current run reliability is '90.47619047619048'. drift is 7.70563 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                 | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69127#01981520-4813-476f-9434-221ee3d1efe4 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69201">build#69201</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite       | test_replica_pair_frequency          |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/69201#0198194c-43d5-4769-86ab-de9cffbc14cb | FAIL          | 0/1      |                                                                                                                                                           |
| DataMigrationsApiTest | test_creating_and_listing_migrations |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69201#01981968-73cb-41e3-a72b-f2971405bfe8 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69351">build#69351</a></summary>


| test_class            | test_method                               | test_arguments                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:----------------------|:------------------------------------------|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePoolMigrationTest | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69351#01982d99-c01f-4bb4-b4c3-9608ef03e748 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69510">build#69510</a></summary>


| test_class                | test_method   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/69510#01983749-3835-443f-afb1-5803f5d3d76e | FLAKY         | 20/21    | upstream reliability is '95.71045576407506'. current run reliability is '95.23809523809523'. drift is 0.47236 and the allowed drift is set to 50. The test should PASS |
| TopicRecoveryTest         | test_fast2    | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69510#0198374c-cb14-4e73-bee3-0732325d2103 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69667">build#69667</a></summary>


| test_class                 | test_method                    | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------|:-------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupBalancingTest | test_coordinator_nodes_balance | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/69667#019841f3-41c4-4f03-ad78-b6b200e5b72f | FLAKY         | 20/21    | upstream reliability is '99.15014164305948'. current run reliability is '95.23809523809523'. drift is 3.91205 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest   | test_node_operations           | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/69667#019841f5-4fee-4d8e-acf6-b1d9a464366b | FLAKY         | 18/21    | upstream reliability is '99.70326409495549'. current run reliability is '85.71428571428571'. drift is 13.98898 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69754">build#69754</a></summary>


| test_class                       | test_method                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------------|:---------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest        | test_write                 | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69754#01985152-3b6d-4105-807c-96c68988cbca | FLAKY         | 20/21    | upstream reliability is '94.83282674772036'. current run reliability is '95.23809523809523'. drift is -0.40527 and the allowed drift is set to 50. The test should PASS |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69754#01985151-34c8-4148-a574-7017026d1902 | FLAKY         | 20/21    | upstream reliability is '99.72451790633609'. current run reliability is '95.23809523809523'. drift is 4.48642 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

michael-redpanda commented:
Force push:

* Rebased off dev to fix merge conflicts

michael-redpanda commented:
Force push:

* Fixed rebase snafu

michael-redpanda committed:
cl: Wire in link instance to task creation

When a task is created, give it a pointer back to the owning link so it
can be referenced by the task.  Will be helpful for when tasks need
access to the remote_cluster_connection instance.

Signed-off-by: Michael Boquard <michael@redpanda.com>

oleiman commented on src/v/kafka/client/test/cluster_mock.h:
```diff
@@ -65,6 +66,23 @@ struct broker_mock_factory : public kafka::client::broker_factory {
+};
+
+struct topic_metadata {
+    int32_t authorized_operations
```
nitpick: i agree with the bot. i had to look this up.

oleiman commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -468,6 +468,28 @@ errc frontend::validator::validate_connection_config(
+        return error_code;
+    }
+
+    if (config.client_id.empty()) {
+        vlog(cluster::clusterlog.info, "Client ID cannot be empty");
+        return error_code;
+    }
```
sorry if obvious, but why is client ID required? not suggesting it shouldn't be, just might be nice to reflect that in the comments or commit message or somewhere.

mmaslankaprv replied:
I would say the system should generate a client id based on the link configuration in a deterministic way. This way we could identify logs related with the link

michael-redpanda replied:
Yeah I'm happy to do either way. We just want to be able to tell the source cluster what the client-id is of the cluster link.  I was thinking if we allow the user to set it, then they can configure the source cluster to tell it what the client-id is of the connection.

michael-redpanda replied:
Otherwise either the customer needs to figure out the client-id and then set it, or we need to 'tell' the source cluster what the client-id is

michael-redpanda replied:
Going to pull this validation stuff out for the time being.  We'll have a separate discussion on client-id

bharathv replied:
> I would say the system should generate a client id based on the link configuration in a deterministic way. This way we could identify logs related with the link

+1

oleiman commented on src/v/cluster/cluster_link/tests/frontend_validation_test.cc:
```diff
@@ -217,15 +225,17 @@ TEST_F_CORO(frontend_validation_test, update_existing_good_uuid) {
       .connection = connection_config{
-        .bootstrap_servers = {net::unresolved_address{"localhost", 9092}}}};
+        .bootstrap_servers = {net::unresolved_address{"localhost", 9092}},
+        .client_id = "client-id"}};
```
nitpick: since there's not much variation here, we could have a default config static on the fixture

oleiman commented on src/v/cluster/cluster_link/tests/frontend_validation_test.cc:
```diff
@@ -294,7 +308,46 @@ TEST_F_CORO(
+      .connection = connection_config{
+        .bootstrap_servers = {net::unresolved_address{"localhost", 9092}},
+        .client_id
+        = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
```
nitpick

```suggestion
        = std::string(65, 'A'),
```

michael-redpanda replied:
booooo

oleiman replied:
don't you mean "booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo"?

michael-redpanda replied:
`"b" + std::string(65, 'o')`

oleiman commented on src/v/cluster/cluster_link/tests/frontend_validation_test.cc:
```diff
@@ -294,7 +308,46 @@ TEST_F_CORO(
+
+TEST_F_CORO(frontend_validation_test, client_id_too_long) {
+    metadata m{
+      .name = name_t("link1\x0d"), // Contains a control character
```
suggestion: test case should probably only have one invalid field.

oleiman commented on src/v/cluster_link/deps.cc:
```diff
@@ -0,0 +1,140 @@
+          cllog.warn, "Exception in remote cluster connection: {}", e.what());
+    }
+}
+} // namespace
```
I don't quite understand why we need this. there's only one usage, and it seems like having the log at the call site would make things easier to track down.

could use `coroutine::as_future` if the scope nesting feels unwieldy. idk.

oleiman commented on src/v/cluster_link/deps.cc:
```diff
@@ -0,0 +1,241 @@
+                "Failed to connect to remote cluster: {}", e.what()));
+        }
+
+        try {
```
suggestion bordering on nitpick: consider `coroutine::as_future` for readability?

oleiman commented on src/v/cluster_link/deps.cc:
```diff
@@ -0,0 +1,241 @@
+            res = err_info(
+              errc::remote_cluster_does_not_support_required_api,
+              ssx::sformat(
+                "Remote cluster does not support describe configs API: {}",
```
q: if `supported_api_versions` throws, does that precisely imply the remote doesn't support describe configs?

oleiman commented on src/v/cluster_link/deps.cc:
```diff
@@ -0,0 +1,241 @@
+            ss::visit(
+              link_connection.cert.value(),
+              [&](const model::tls_file_path& path) {
+                  vassert(
+                    std::holds_alternative<model::tls_file_path>(
+                      link_connection.key.value()),
```
^^ this (**Copilot's suggestion**) is phrased stupidly, but it might be a bit neater w/ a visitor that takes multiple variant instances

like this: 

```c++
struct visitor {
    auto operator()(model::tls_file_path&, model::tls_file_path&) {
        // OK!
        return stuff;
    }
    ....
    template<typename T1, typename T2>
    requires(!std::is_same_v<T1, T2>)
    auto operator()(T1&, T2&) {
        vassert(false, "OH NO!");
    }
};
```

then you can call it like 

```c++
auto my_stuff = std::visit(visitor{}, connection.cert.value(), connection.key.value());
```

oleiman replied:
my other question is whether this should actually assert. where is the sameness of the key & cert variant types enforced, and if violated does that indicate a logic error in the code?

michael-redpanda replied:
> where is the sameness of the key & cert variant types enforced

in the frontend validation code

oleiman commented on src/v/cluster_link/manager.cc:
```diff
@@ -130,22 +130,17 @@ ss::future<> manager::handle_on_link_change(model::id_t id) {
-                  id);
                 try {
-                    auto ec = co_await new_link->register_task(std::move(task));
+                    auto ec = co_await new_link->register_task(
+                      task_factory.get());
```
seems unlikely, but is it possible for the task factory to become invalid across this scheduling point?

mmaslankaprv replied:
yeah, there is not protection from changing the `_task_factories` collection while iterating over them here

oleiman reviewed (commented):
all looks pretty sensible to me. handful of nits and questions.

mmaslankaprv commented on src/v/kafka/client/topic_cache.h:
```diff
@@ -29,6 +29,9 @@ class topic_cache {
 
     struct topic_data {
         chunked_hash_map<model::partition_id, partition_data> partitions;
+        int32_t authorized_operations
```
Do we already have a named type for this ? If no then i think it would be worth introducing one

michael-redpanda replied:
No we do not.  I can introduce that in kafka/protocol

oleiman replied:
since it's a bitfield and has a meaningful default value, might be worth a whole struct. my $0.02

mmaslankaprv commented on src/v/kafka/client/topic_cache.h:
```diff
@@ -49,6 +52,13 @@ class topic_cache {
     std::optional<kafka::leader_epoch>
       leader_epoch(model::topic_partition_view) const;
 
+    chunked_vector<model::topic> topics() const;
```
why do we need a copy ? It would be better to return `topics_t` const ref

mmaslankaprv replied:
Also copilot suggest using view, it is also an option here

mmaslankaprv commented on src/v/cluster_link/model/types.h:
```diff
@@ -118,23 +123,21 @@ struct connection_config
+    /// The CA file to use
+    std::optional<tls_file_or_value> ca;
+    /// The client ID to use
+    ss::sstring client_id;
```
This could be autogenerated by the cluster linking subsystem f.e. based on target cluster id. This way we will be able to identify the client logs related with a certain link

michael-redpanda replied:
I was thinking of using the cluster link name as part of the client id

mmaslankaprv commented on src/v/cluster_link/model/types.h:
```diff
@@ -105,6 +105,11 @@ struct scram_credentials
     auto serde_fields() { return std::tie(username, password, mechanism); }
 };
 
+using tls_file_path = named_type<ss::sstring, struct tls_file_path_tag>;
```
do we even need paths ?

michael-redpanda replied:
I guess depends on how the customer wants to deploy the solution... the option is there

michael-redpanda replied:
Seems that cloud may still want paths for TLS config

mmaslankaprv commented on src/v/cluster_link/deps.h:
```diff
@@ -0,0 +1,129 @@
+    virtual ss::future<> request_metadata_update() = 0;
+    virtual bool is_connected() const = 0;
+    virtual ss::future<kafka::describe_configs_response> describe_topics(
+      chunked_vector<::model::topic> topics,
+      chunked_vector<ss::sstring> config_keys = {})
```
const ref ?

mmaslankaprv replied:
or move

mmaslankaprv reviewed (commented):
overall looks good, left some comments

michael-redpanda commented:
@mmaslankaprv when I was working on my next task, I discovered that you already kind of built in a 'mock' system within `cluster` that allows us to provide 'mocked' `brokers`.  I think that's a better system than the `remote_cluster_connection` I did in here.  I'm going to make some changes to use that instead.

michael-redpanda committed:
cl: Added utility to create k/client config

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Introduce cluster_factory

Used to create instances of kafka::client::cluster

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/tests: Wire in cluster_mock

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire in cluster_factory into manager

Signed-off-by: Michael Boquard <michael@redpanda.com>

bharathv commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -451,4 +439,35 @@ errc frontend::validator::validate_mutation(const cluster_link_cmd& cmd) const {
+  cluster::cluster_link::errc error_code) const {
+    if (config.bootstrap_servers.empty()) {
+        vlog(
+          cluster::clusterlog.info,
```
warn?

oleiman replied:
bump (and same comment for the other `INFO` logs in this function)

michael-redpanda replied:
Sorry missed this.  I've seen other validators use info level.  Is warn too high for a user level mistake for config?  I'm not arguing against fixing it just posting questions.  :)

michael-redpanda replied:
Sorry missed this.  I've seen other validators use info level.  Is warn too high for a user level mistake for config?  I'm not arguing against fixing it just posting questions.  :)

bharathv commented on src/v/cluster_link/deps.h:
```diff
@@ -0,0 +1,63 @@
+    virtual ~link_factory() = default;
+
+    virtual std::unique_ptr<link> create_link(
+      ::model::node_id self,
```
nit: curious why ::model..

michael-redpanda replied:
Because there's a `cluster_link::model` namespace

bharathv commented on src/v/cluster_link/deps.cc:
```diff
@@ -0,0 +1,140 @@
+
+        if (res.code() != errc::success) {
+            co_await handle_exception([this] { return _cluster.stop(); });
+        }
```
can we ensure that anything that is start()-ed should be stop()-ed too (will avoid these patterns).

bharathv commented on src/v/cluster_link/deps.cc:
```diff
@@ -0,0 +1,140 @@
+
+        try {
+            auto describe_configs_version
+              = co_await _cluster.supported_api_versions(
```
whats the retry story? This may be flaky even if one broker doesn't respond.

michael-redpanda committed:
cl: Wire in cluster to link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Fix possible issue with task factory loop

While looping through the test factories and registering new tasks, it
is possible that the task_factories would be modified while awaiting the
register coroutine.  Lock the task reconciler mutex while registering
new task factories to ensure that doesn't happen.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/protocol: Added named types for authorized ops

Added named types for topic and cluster authorized operations field in
metadata response.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
@oleiman @mmaslankaprv @bharathv this is ready for re-review when you all have a chance

mmaslankaprv commented on src/v/cluster_link/utils.cc:
```diff
@@ -0,0 +1,98 @@
+          link_connection.authn_config.value());
+    }
+
+    config.client_id = link_connection.client_id;
```
Is this `link_connection.client_id` autogenerated ? It is important to make sure it is unique as otherwise the client metrics will clash with each other

michael-redpanda replied:
I believe that's the direction we are going to go in, but I want to have a discussion about how to get the source cluster said cluster-id.  It will be set in a following PR

michael-redpanda replied:
per discussion we will autogenerate - i'll do it in a follow up

oleiman commented on src/v/kafka/client/test/cluster_test_with_mock.cc:
```diff
@@ -134,3 +134,30 @@ TEST_F(cluster_mock_fixture, TestApiVersionDiscovery) {
+    const auto& topics = cluster.get_topics();
+    auto num_parts = topics.partition_count(model::topic_view{"test-topic"});
+    ASSERT_TRUE(num_parts.has_value());
+    EXPECT_EQ(num_parts.value(), 3);
```
fyi: IIRC, `EXPECT_EQ` should deal with unpacking `std::optional` on your behalf. i.e. no `has_value` required. 

```suggestion
    EXPECT_EQ(num_parts, 3);
```

oleiman commented on src/v/cluster_link/model/types.h:
```diff
@@ -118,23 +123,18 @@ struct connection_config
-    ss::sstring key_file_path;
-    /// Path to the CA file to use
-    ss::sstring ca_file_path;
+    /// certificate file to use
+    std::optional<tls_file_or_value> cert;
+    /// key to use (when mTLS is in use)
+    std::optional<tls_file_or_value> key;
+    /// The CA file to use
+    std::optional<tls_file_or_value> ca;
```
Are we confident that `serde::variant` behaves sanely w/ `std::optional`? Doesn't seem like it should cause problems, but the implementation is sort of wild iirc.

michael-redpanda replied:
Oh I haven't heard of any issues with this...  I did this because monostate wasn't supported in serde

oleiman replied:
ya, i don't rightly know. there's some wacky stuff with versioning, but that's not an issue here. I want to say there's an example usage in transforms that has `nullopt_t` as a variant alternative, but idk why that was done. just flagging as kind of a fussy construct.

oleiman commented on src/v/cluster_link/model/types.h:
```diff
@@ -335,6 +335,60 @@ struct fmt::formatter<
+        auto it = ctx.begin();
+        auto end = ctx.end();
+
+        if (it != end && (*it == 's')) {
+            _is_sensitive = true;
+            ++it;
+        }
```
nitpick: idk what the 's' means here...can we add a comment or something?

michael-redpanda replied:
S for super sensitive secret

oleiman replied:
🤫

oleiman commented on src/v/cluster_link/tests/deps.h:
```diff
@@ -154,6 +156,21 @@ class fake_partition_leader_cache
+    explicit cluster_mock_factory(kafka::client::cluster_mock* cluster_mock)
+      : _cluster_mock(cluster_mock) {}
+
+    kafka::client::cluster create_cluster(const model::metadata& md) override {
```
nitpick: `final`

oleiman commented on src/v/cluster_link/manager.h:
```diff
@@ -62,7 +62,13 @@ class manager {
+        auto fut = co_await ss::coroutine::as_future(
+          _link_task_reconciler_mutex.get_units(_as));
+        if (fut.failed()) {
+            // abort source triggered, exit early
+            co_return;
```
issue: you need to call `fut.get_exception` on this path to avoid abandoned future exception

michael-redpanda replied:
Oh shoot thanks

oleiman commented on src/v/cluster_link/link.cc:
```diff
@@ -137,53 +137,20 @@ ss::future<> link::stop() {
+      t,
+      "Failed to create task from factory {} for cluster link {}",
+      tf->created_task_name(),
+      _config.name);
```
question: is this really a crash-worthy condition? maybe some color in a code comment would help.

michael-redpanda replied:
I'll add an error code and return it

oleiman reviewed (commented):
pretty much lgtm. some nits and one real issue (small change).

oleiman reviewed (approved):
lgtm

dotnwat commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -437,18 +421,49 @@ errc frontend::validator::validate_mutation(const cluster_link_cmd& cmd) const {
+      && config.cert.value().index() != config.key.value().index()) {
+        vlog(
+          cluster::clusterlog.warn,
+          "If providing a certificate or key, both must be file paths or "
+          "both must be values");
```
just wondering does this error message match the situation (i haven't dug yet into what is inside `index()`)

michael-redpanda replied:
I do have a unit test that validates this condition
